### ISSUE TITLE
timezone and certification expired

### ISSUE THREAD

**Issue by vanloswang** [openshift/origin#11089]
##### Version

Origin 1.2
##### Steps To Reproduce
1. Choose timezone to Asia/Shanghai install CentOS 7 minimal in a VM, the host machine timezone is Asia/Shanghai too.
2. use openshift-ansible to install Origin 1.2
3. reboot the VM
##### Current Result

The openshift-master is not started, and report error as following:

``` bash
x509: certificate has expired or is not yet valid
```

Then set the clock to 8 hours(The UTC and CST is differ 8 hours) later in the VM and restart openshift-master and openshift-node service, it will start OK.
##### Expected Result

start ok.

This is an issue about certification generation, I guess. But after all, I did not find some clue in openshift-ansible or opensfhit origin source code. Does someone meet the same issue or have some suggestions?

**Comment by detiber** [comment]
@vanloswang Can you provide more information on how you installed:
- did you use the quick installer? 
- did you use the advanced install? If so, can you provide your hosts file?

Can you provide more of the error output? In some cases we generate the etcd certificates differently than we generate the other certificates in use by the master.

Also, any more information you can provide about the timezone config of the server and the generated certificates would also be helpful.
- Output of `timedatectl` on the system should be sufficient for timezone config of the server
- Output of `openssl x509 -in <cert file> -text`, particularly the 'Validity section' and the approximate time the system was installed in the configured timezone.

**Comment by vanloswang** [comment]
I use quick install to generate installer.cfg.yml and stop the install process, then modify it and use unattached install. The installer.cfg.yml is as following:
```bash
# cat ~/.config/openshift/installer.cfg.yml 
ansible_config: /usr/share/atomic-openshift-utils/ansible.cfg
ansible_log_path: /tmp/ansible.log
ansible_ssh_user: root
hosts:
- connect_to: 192.168.1.217
  hostname: 192.168.1.217
  ip: 192.168.1.217
  master: true
  node: true
  public_hostname: 192.168.1.217
  public_ip: 192.168.1.217
  storage: true
variant: origin
variant_version: '1.2'
version: v1
```

Now the time info in the VM is as following:
```bash
# timedatectl
      Local time: 二 2016-09-27 17:16:00 CST
  Universal time: 二 2016-09-27 09:16:00 UTC
        RTC time: 二 2016-09-27 09:16:00
       Time zone: Asia/Shanghai (CST, +0800)
     NTP enabled: yes
NTP synchronized: no
 RTC in local TZ: no
      DST active: n/a
```

The validity timezone is GMT. full info is as following:
```bash
# openssl x509 -in ca.crt -text                
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 1 (0x1)
    Signature Algorithm: sha256WithRSAEncryption
        Issuer: CN=openshift-signer@1474911179
        Validity
            Not Before: Sep 26 17:33:01 2016 GMT
            Not After : Sep 25 17:33:02 2021 GMT
```

**Comment by vanloswang** [comment]
The etcd related crt validity is as following:
```bash
# openssl x509 -in etcd.server.crt -text     
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 7 (0x7)
    Signature Algorithm: sha256WithRSAEncryption
        Issuer: CN=openshift-signer@1474911179
        Validity
            Not Before: Sep 26 17:33:02 2016 GMT
            Not After : Sep 26 17:33:03 2018 GMT
# openssl x509 -in master.etcd-client.crt -text
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 2 (0x2)
    Signature Algorithm: sha256WithRSAEncryption
        Issuer: CN=openshift-signer@1474911179
        Validity
            Not Before: Sep 26 17:33:01 2016 GMT
            Not After : Sep 26 17:33:02 2018 GMT
```

**Comment by detiber** [comment]
@vanloswang thanks for the additional info. It definitely looks like the openshift certificate tooling is not properly converting the local time to UTC prior to configuring the validity time.

@liggitt I believe https://github.com/openshift/origin/pull/11107 addresses the issue causing this.

**Comment by liggitt** [comment]
I don't mind converting to UTC, but changing time zones doesn't do anything to the absolute epoch-relative timestamp... I'm not sure how that would affect validity

**Comment by detiber** [comment]
@liggitt I don't see where an epoch relative timestamp comes into play. time.Now() returns the current local time (https://golang.org/pkg/time/#Now) and TLS validity timestamps are specified in UTC or Generalized time (https://www.ietf.org/rfc/rfc3280.txt).

**Comment by liggitt** [comment]
the golang lib [converts NotBefore/NotAfter to UTC already](https://github.com/golang/go/blob/release-branch.go1.6/src/crypto/x509/x509.go#L1609). `openssl x509` shows no difference in the timestamps in the generated certs with your change :-/

**Comment by liggitt** [comment]
the times [here](https://github.com/openshift/origin/issues/11089#issuecomment-249743140) seem odd:

Now the time info in the VM is as following:
```bash
# timedatectl
      Local time: 二 2016-09-27 17:16:00 CST
  Universal time: 二 2016-09-27 09:16:00 UTC
        RTC time: 二 2016-09-27 09:16:00
       Time zone: Asia/Shanghai (CST, +0800)
     NTP enabled: yes
NTP synchronized: no
 RTC in local TZ: no
      DST active: n/a
```

The validity timezone is GMT. full info is as following:
```bash
# openssl x509 -in ca.crt -text                
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 1 (0x1)
    Signature Algorithm: sha256WithRSAEncryption
        Issuer: CN=openshift-signer@1474911179
        Validity
            Not Before: Sep 26 17:33:01 2016 GMT
            Not After : Sep 25 17:33:02 2021 GMT
```

These two times should both be UTC/GMT, but they don't agree at all, right?

Universal time: 二 2016-09-27 09:16:00 UTC
Not Before: Sep 26 17:33:01 2016 GMT

**Comment by detiber** [comment]
Ah, I missed the automatic conversion and I missed the date difference (26th vs 27th).

**Comment by vanloswang** [comment]
@liggitt @detiber timedatectl is execute today, but the cert is generated yesterday. So there is 26th vs 27th. The timezone is GMT in  cert validity section confused me,  Is it should automatic convert to UTC?

**Comment by detiber** [comment]
@vanloswang GMT is covered under the Generalized time part of RFC 3280, so that is expected. I will try and replicate the issue locally to see if I can determine where things are going off the rails.

**Comment by detiber** [comment]
Okay, so I think I've discovered a problem (whether or not it is *the* problem is another story). It looks like we are currently configuring ntp/chrony after we generate the certs, which could definitely lead to a similar issue as the one reported.

**Comment by vanloswang** [comment]
@detiber @liggitt I meet a same cert error when enabling cluster metrics. I deploy it by using self signed certs and the certs is auto generated by origin. The metrics-deployer container always said following information to me:
```
Creating the Cassandra Certificate Secrets configuration json file
+++ base64 -w 0 /etc/deploy/_output/hawkular-cassandra.cert
+++ base64 -w 0 /etc/deploy/_output/hawkular-cassandra-ca.cert
++ echo 'Creating Hawkular Metrics & Cassandra Secrets'
++ oc create -f /etc/deploy/_output/hawkular-metrics-secrets.json
Creating Hawkular Metrics & Cassandra Secrets
unable to connect to a server to handle "secrets": Get https://192.168.10.16:443/api: x509: certificate has expired or is not yet valid
```

**Comment by vanloswang** [comment]
@detiber @liggitt metrics deployer works ok now, the url is wrong, it should use 8443 instead of 443.

**Comment by abdelhegazi** [comment]
Hi there,

I have got the same issue again.
I am running origin 1.3 which originally installed via ansbile-playbook 
I followed up all mentioned solutions but I still have got the api server isn't available due to the certificate expiry.

I would like to understand why this happened, I checked all certificate under /etc/origin/master as well as on all nodes and non of them seem to be expired by all ways.

I thought I may be able to generate these certificate somewhere and just copy it but I dont see any reason for that,

Any suggestions ?
